In [81]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

# Carregando os dados
data = pd.read_csv('sprint-chatbot.csv', sep=';')
data = data.dropna(axis=1)

# Pré-processamento dos dados
le = LabelEncoder()
data['Marca da Bike'] = le.fit_transform(data['Marca da Bike'])
data['Modelo da Bike'] = le.fit_transform(data['Modelo da Bike'])
data['Tipo da bicicleta'] = le.fit_transform(data['Tipo da bicicleta'])


In [82]:
data.head()

Marca da Bike  Modelo da Bike  Tipo da bicicleta         Valor
0              8              36                  2  R$ 35.999,00
1             29              87                  9  R$ 21.490,00
2              9              57                  2  R$ 26.706,63
3              4              17                  2  R$ 26.999,00
4             16              58                  2  R$ 26.990,00

In [83]:
data.head(35)

Marca da Bike  Modelo da Bike  Tipo da bicicleta          Valor
0               8              36                  2   R$ 35.999,00
1              29              87                  9   R$ 21.490,00
2               9              57                  2   R$ 26.706,63
3               4              17                  2   R$ 26.999,00
4              16              58                  2   R$ 26.990,00
5               2              75                  2   R$ 40.062,61
6               7              91                  2   R$ 32.407,94
7              19              69                  2   R$ 38.999,00
8              23               0                  2   R$ 38.500,00
9              10              68                  2   R$ 27.000,00
10             20              35                  2  R$ 151.200,00
11              1              88                  2   R$ 49.294,55
12             23              82                  2   R$ 16.559,91
13             28              34                  2   R$ 26.680,00
14              6              28                  2   R$ 26.000,00
15             12              76                  2   R$ 11.039,08
16             26             102                  2   R$ 10.990,00
17             22              59                  4    R$ 5.190,00
18             21              77                  7    R$ 1.899,00
19             11              13                  6    R$ 4.499,00
20              3              84                  2    R$ 7.580,00
21             17              81                  8    R$ 6.472,80
22             14              15                  3      R$ 15.000
23             15             103                  0    R$ 3.000,00
24             30              14                  2    R$ 2.024,10
25             13              78                  7    R$ 1.259,10
26             13              80                  7    R$ 1.574,10
27             13              89                  7    R$ 1.979,10
28             13              92                  7    R$ 1.979,10
29             13              79                  7    R$ 1.979,10
30              5              37                  1    R$ 6.990,00
31              5              38                  1   R$ 25.959,00
32              5              39                  1   R$ 32.299,00
33              5              40                  1   R$ 43.999,00
34              5              41                  1    R$ 8.199,00

In [84]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 104 entries, 0 to 103
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Marca da Bike      104 non-null    int64 
 1   Modelo da Bike     104 non-null    int64 
 2   Tipo da bicicleta  104 non-null    int64 
 3   Valor              104 non-null    object
dtypes: int64(3), object(1)
memory usage: 3.4+ KB


In [85]:
# Limpeza e conversão da coluna 'Valor'
data['Valor'] = data['Valor'].str.replace('R\$', '').str.replace('.', '').str.replace(',', '.').astype(float)

# Divisão dos dados em treinamento e teste
X = data.drop('Tipo da bicicleta', axis=1)
y = data['Tipo da bicicleta']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

<ipython-input-85-783576d64e54>:2: FutureWarning: The default value of regex will change from True to False in a future version.
  data['Valor'] = data['Valor'].str.replace('R\$', '').str.replace('.', '').str.replace(',', '.').astype(float)
<ipython-input-85-783576d64e54>:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data['Valor'] = data['Valor'].str.replace('R\$', '').str.replace('.', '').str.replace(',', '.').astype(float)


In [86]:
X_test.head(5)

Marca da Bike  Modelo da Bike    Valor
30              5              37   6990.0
65             18              83   7749.0
64             18              19  17890.0
53              5             100   1799.0
45              5              52   3299.0

In [87]:
X_train.head(5)

Marca da Bike  Modelo da Bike    Valor
55              5              24   1649.0
22             14              15  15000.0
76              0              10  52999.0
44              5              51   3299.0
72              0              99  59999.0

In [88]:
y_train.head(5)

55    10
22     3
76     5
44     6
72     8
Name: Tipo da bicicleta, dtype: int64

In [89]:
from sklearn.preprocessing import MinMaxScaler

#Escalonamento
valor = data['Valor'].values.reshape(-1, 1)

scaler = MinMaxScaler()

valor_scaled = scaler.fit_transform(valor)

In [90]:
# Treinamento do modelo
model = RandomForestClassifier(n_estimators=100, random_state=42)
# Pegando o modelo e treinando (.fit)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [91]:
# Previsões
y_pred = model.predict(X_test)

In [92]:
# Avaliação do modelo
accuracy = accuracy_score(y_test, y_pred)
print(f'Acurácia: {accuracy}')
print(classification_report(y_test, y_pred))

Acurácia: 0.8095238095238095
              precision    recall  f1-score   support

           1       1.00      1.00      1.00         3
           2       0.60      0.75      0.67         4
           5       0.80      1.00      0.89         4
           6       0.75      1.00      0.86         3
           7       1.00      1.00      1.00         2
           8       1.00      0.50      0.67         2
          10       1.00      0.33      0.50         3

    accuracy                           0.81        21
   macro avg       0.88      0.80      0.80        21
weighted avg       0.85      0.81      0.79        21



In [93]:
import pickle

In [94]:
with open('meu_modelo_serializado.pickle', 'wb') as f:
    pickle.dump(model, f)

In [95]:
with open('meu_modelo_serializado.pickle', 'rb') as f:
    modelo_carregado = pickle.load(f)

In [96]:
modelo_carregado.predict([[5,37,6990.0]])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


array([1])

Para fazer o modelo de classificação, pegamos o dataset feito anteriormnete em arquivo csv, possuindo 4 colunas: Marca da Bike, Modelo da Bike, Tipo da Bike e Valor.
Foi utilizado como dados de entrada a marca da bike, modelo da bike e valor, 3 colunas, e com esses dados, foi feito um treinamento com Random Forest Classifier para então a ML prever, baseados nessas 3 colunas, qual é o tipo da bicicleta, a coluna restante. Como os valores que devem ser a entrada devem ser categóricos, foi feito uma conversão com Label Encoder. Foi colocado também o passo de Escalonamento, mesmo não sendo preciso para a nossa ML.Obtivemos uma acurácia de 81%.